### PR TITLE
kep: add arena v2 proposal

### PR DESCRIPTION
**Pull request by vicoooo26**
## Purpose of this PR

This PR introduces the Arena v2 design proposal (KEP), documenting a complete architectural redesign of Arena to eliminate Helm dependencies, reduce code complexity, and introduce YAML-first configuration.

**Proposed changes:**

- Add Arena v2 KEP (`keps/arena-v2/README.md`) covering motivation, architecture overview, design principles, user stories, test plan, and graduation criteria
- Add YAML schema specification (`keps/arena-v2/yaml-schema.md`) defining the structured job configuration format for v2
- Add CLI flag mapping (`keps/arena-v2/cli-flag-mapping.md`) documenting v1-to-v2 flag compatibility for the `arena submit` legacy interface
- Add `CLAUDE.md` with project conventions and build instructions for AI-assisted development

## Change Category

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that could affect existing functionality)
- [x] Documentation update

### Rationale

Arena v1 suffers from architectural limitations: a heavy Helm dependency (24 charts + Helm Go SDK), a 6-step submission pipeline with temp file/ConfigMap leak risks, flag-only configuration (40+ flags per job with no versioning), and high code complexity (thousands of lines spread across `pkg/commands/`, `pkg/training/`, `pkg/argsbuilder/`, `pkg/apis/`).

Arena v2 addresses these by:
1. **Generating CRDs directly in Go** using `unstructured.Unstructured` — no Helm, no shell-outs to `kubectl`
2. **Introducing YAML-first configuration** — users can version, review, and reuse job configs as code
3. **Adopting a modular package structure** — `pkg/cli`, `pkg/task`, `pkg/provider`, `pkg/client`, `pkg/output` with clean separation of concerns
4. **Providing backward compatibility** — the `arena submit <type> --flags` interface maps v1 flags to v2's internal model

This PR contains the design documentation only. Implementation will follow in subsequent PRs.

### CODE REVIEW

File: keps/arena-v2/cli-flag-mapping.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+| `--success-policy` | `lifecycle.success_policy` | `--success-policy` | ✅ (moved to lifecycle block) |
+| `--role-sequence` | — | — | ❌ Implementation detail, not user-facing |
+
+### 6b. PyTorchJob
+
+| v1 Flag | v2 YAML | v2 CLI Flag | Status |
+|---------|---------|-------------|--------|
+| `--cpu` | `resources.cpus` | `--cpus` | ✅ |
+| `--memory` | `resources.memory` | `--mem` | ✅ |
+| `--nproc-per-node` | `framework.pytorch.policy.nproc_per_node` | `--nproc-per-node` | ✅ (`auto` / `gpu` / `cpu` / int) |
```
Comment (cheyang) [inline keps/arena-v2/cli-flag-mapping.md]: The YAML path here doesn't match `yaml-schema.md`. The schema documents the object form as `framework: { name: pytorch, options: { nproc_per_node: ... } }` and its own override table maps `--nproc-per-node` to `framework.options.nproc_per_node`. This file uses `framework.pytorch.policy.nproc_per_node` instead, and the same pattern repeats for `framework.tensorflow.policy.ps_count`, `framework.mpi.policy.gpu_topology`, `framework.horovod.policy.ssh_port`, etc. The two documents describe two different schemas — please pick one and use it everywhere.

File: keps/arena-v2/cli-flag-mapping.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+| `--name` | string | required | `name` | `--name` | ✅ |
+| `--image` | string | required | `image` | `--image` | ✅ |
+| `--image-pull-policy` | string | `"Always"` | `image_pull_policy` | `--image-pull-policy` | ✅ |
+| `--image-pull-secret` | string[] | `[]` | `image_pull_secrets` | `--image-pull-secret` | ✅ (reference-only, no auto-create) |
+
+### 2b. Resources
+
+| v1 Flag | Type | Default | v2 YAML | v2 CLI Flag | Status |
+|---------|------|---------|---------|-------------|--------|
+| `--gpus` | int | `0` | `resources.nvidia.com/gpu` | `--gpus` | ✅ |
```
Comment (cheyang) [inline keps/arena-v2/cli-flag-mapping.md]: Resources here are written as `resources.nvidia.com/gpu` / `resources.cpus` / `resources.memory`, but `yaml-schema.md` puts resources under `worker.resources` (the section is explicitly titled `worker.resources` and the example uses `worker.resources.nvidia.com/gpu`). The mapping needs the `worker.` prefix, or the schema needs a top-level `resources` block — currently neither aligns with the other.

File: keps/arena-v2/yaml-schema.md [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+
+### resources
+
+Fields under `worker.resources` are all **Scalar** values, applied identically to both `requests` and `limits` (**Guaranteed QoS**).
+
+```yaml
+worker:
+  resources:
+    nvidia.com/gpu: 8
+    cpu: 32
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md:185]: The example here uses singular `cpu: 32`, but the CLI override table further down and the entire `cli-flag-mapping.md` map `--cpus` to `resources.cpus` (plural). K8s itself uses `cpu`; `cpus` is unusual. Worth picking one spelling and aligning all three documents.

File: keps/arena-v2/cli-flag-mapping.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+| `--gpu-type` | string | `""` | (node selector hint) | `--gpu-type` | ✅ Maps to `nvidia.com/gpu.product` label |
+| `--cpu` | string | `""` | `resources.cpus` | `--cpus` | ✅ |
+| `--memory` | string | `""` | `resources.memory` | `--mem` | ✅ |
+| `--device` | string[] | `[]` | `resources.<name>` (flat) | `--device` | ✅ e.g. `--device hugepages-2Mi=32Gi` |
+| `--workers` | int | `1` | `workers` | `--workers` | ✅ |
```
Comment (cheyang) [inline keps/arena-v2/cli-flag-mapping.md]: `--workers` is mapped to a top-level `workers` field that doesn't exist in the schema. The correct target is `worker.replicas` (which is what `yaml-schema.md`'s own override table says).

File: keps/arena-v2/cli-flag-mapping.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+---
+
+## 6. Framework-Specific Flags
+
+### 6a. TFJob
+
+| v1 Flag | v2 YAML | v2 CLI Flag | Status |
+|---------|---------|-------------|--------|
+| `--ps` (count) | `framework.tensorflow.policy.ps_count` | `--ps-count` | ✅ |
+| `--ps-image` | `roles[]` (name: ps, independent declaration) | — | ✅ YAML-only |
```
Comment (cheyang) [inline keps/arena-v2/cli-flag-mapping.md]: `roles[]` is referenced throughout sections 6a and 6e (`roles[] (name: ps, ...)`, `roles[] (name: chief, ...)`, `roles[] (name: worker, ...)`, `roles[] (name: evaluator, ...)`, `roles[] (name: launcher, ...)`) but the schema never defines a `roles[]` array — only a flat `worker` block exists. Either add the `roles[]` (or equivalent per-role) construct to `yaml-schema.md`, or rewrite these rows to describe how per-role overrides actually work. Without it, TFJob (ps/chief/evaluator) and DeepSpeed/ETJob (launcher) can't be configured at all in v2.

File: keps/arena-v2/yaml-schema.md [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+Init containers generated by `sync` use exec form (CLI auto-generates the command without shell wrapping) — no `run`/`shell` fields needed.
+
+Init containers do not support `envs`, `secrets`, or `resources` configuration.
+
+`sync` and `init` can coexist. `sync` entries are processed before `init` entries.
+
+### worker.replicas → Provider Mapping
+
+```
+User writes: worker.replicas: 4
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md:463]: The semantics of `worker.replicas` are not consistent across frameworks here. Given the same `worker.replicas: 4`:
- pytorch → Master(1) + Worker(3) = 4 pods   (replicas counts Master)
- tensorflow → Chief(1) + Worker(4) = 5 pods (replicas does NOT count Chief)
- mpi → Launcher(1) + Worker(4) = 5 pods    (replicas does NOT count Launcher)

The same number means "total GPU pods" for PyTorch but "worker-only count" for TF/MPI. That's going to surprise users porting jobs between frameworks. Suggest normalizing on one rule (probably: `worker.replicas` is always the worker count, Chief/Launcher/Master are independent) or at minimum calling this discrepancy out in the schema text, not only in this mapping block.

File: keps/arena-v2/cli-flag-mapping.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+### 6e. DeepSpeed
+
+| v1 Flag | v2 YAML | v2 CLI Flag | Status |
+|---------|---------|-------------|--------|
+| `--cpu` | `resources.cpus` | `--cpus` | ✅ |
+| `--memory` | `resources.memory` | `--mem` | ✅ |
+| `--launcher-selector` | `roles[]` (name: launcher, `scheduling.node_selector`) | — | ✅ YAML-only |
+| `--job-restart-policy` | `restart` | `--restart` | ✅ |
+| `--job-backoff-limit` | `lifecycle.backoff_limit` | `--backoff-limit` | ✅ |
+| `--ssh-secret` | — | — | ❌ NOT IN SCHEMA (no ssh_secret in framework.deepspeed) |
```
Comment (cheyang) [inline keps/arena-v2/cli-flag-mapping.md]: `--ssh-secret` is marked `❌ NOT IN SCHEMA` for both DeepSpeed and ETJob without explanation. MPI/DeepSpeed traditionally need an SSH secret for launcher↔worker. If the assumption is that the newer MPI operator's built-in SSH key generation makes this unnecessary, please state that explicitly so existing users can tell whether their workload still works on v2.

File: keps/arena-v2/cli-flag-mapping.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+| `--hostPID` | bool | `false` | `host_pid` | `--host-pid` | ✅ (top-level runtime field, not under scheduling) |
+| `--queue` | string | `""` | `scheduling.queue` | — | ✅ (YAML field exists) |
+| `--rdma` | bool | `false` | — | — | ❌ NOT PLANNED |
+
+### 2d. Data and Volumes
+
+| v1 Flag | Type | Default | v2 YAML | v2 CLI Flag | Status |
+|---------|------|---------|---------|-------------|--------|
+| `-d, --data` | string[] | `[]` | `storages[].pvc` + `storages[].mount_path` | `-d, --data` | ✅ |
+| `--data-dir` | string[] | `[]` | — | — | ❌ hostPath mounts not in schema |
```
Comment (cheyang) [inline keps/arena-v2/cli-flag-mapping.md]: Dropping `--data-dir` (hostPath) and `--config-file` (ConfigMap mount) is in tension with the README's claim of backward compatibility via legacy `arena submit`. The schema already supports `storages[].hostpath`, so `--data-dir` could in principle map to it — worth either documenting that lowering, or softening the no-breaking-change claim.

File: keps/arena-v2/yaml-schema.md [hunk 9]
```diff
@@ -0,0 +1,619 @@
...
+`sync` supports three source types:
+
+| Type | Source | Init Container |
+|---|---|---|
+| `git` | Git repository URL | `alpine/git` clone |
+| `rsync` | Local path (synced at submit time) | `busybox` + ConfigMap/tar injection |
+| `hdfs` | HDFS path | `hadoop/hadoop` init container with `hdfs get` |
+
+```yaml
+sync:
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md:417]: `rsync` is described as "CLI tars + injects via ConfigMap or PVC at submit time". ConfigMaps have a ~1 MiB etcd object limit which datasets and even source trees regularly exceed. Worth specifying when the PVC path is used vs. ConfigMap, and how the PVC is provisioned, or capping ConfigMap-mode payloads with a clear error.

File: keps/arena-v2/yaml-schema.md [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+  scheduler_name: default              # Custom scheduler (optional)
+  queue: low-priority                  # Queue name (optional)
+  node_selector:
+    disktype: ssd
+    zone: us-west-2a
+  tolerations:
+    - key: nvidia.com/gpu
+      operator: Exists
+      effect: NoSchedule
+  affinity:                          # future feature
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md:115]: `scheduling.affinity` is marked `# future feature` in the schema block, but it's still in the CLI override tables without that caveat. Either move it out of the MVP schema or label the corresponding CLI flag rows as Phase 2 so readers don't expect alpha support.

File: keps/arena-v2/yaml-schema.md [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+annotations:                         # Optional
+  note: "experiment run"
+
+image: nvcr.io/nvidia/pytorch:23.10  # Required. Container image.
+run: torchrun train.py --epochs 10   # Required. Training command, executed via shell.
+shell: /bin/sh                       # Optional. Shell interpreter path. Default: /bin/sh. Invalid values fall back to default.
+working_dir: /root                   # Optional. Container working directory (default: image default)
+
+# ─── Task ───
+framework:                           # Required. String shorthand or object with provider config.
```
Comment (cheyang) [inline keps/arena-v2/yaml-schema.md:40]: The comment says `String shorthand or object with provider config`, but only the object form is documented in the "framework (Provider Selection & Config)" section and only the object form is shown in examples. Either implement and document the `framework: pytorch` shorthand, or drop "String shorthand" from this comment.

File: keps/arena-v2/README.md [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+
+As a platform engineer, I want to add support for a new training framework (e.g., DeepSpeed) by implementing a single Go file rather than creating a Helm chart, argsbuilder, command file, and type definitions across 5+ files.
+
+**Story 3: Operator maintaining backward compatibility**
+
+As an operator maintaining existing CI/CD pipelines, I want to continue using `arena submit tfjob --name my-job --workers 4 --gpus 2 "python train.py"` without rewriting my automation scripts, while the platform team gradually adopts v2's YAML-based workflows.
+
+### Notes/Constraints/Caveats
+
+- v2 is a totally new project that live in `v2` branch and build from scratch, it does not import any v1 packages.
```
Comment (cheyang) [inline keps/arena-v2/README.md]: Small grammar nit: "v2 is a totally new project that live in `v2` branch and build from scratch" reads better as "v2 is a separate project that lives in the `v2` branch and is built from scratch".

File: keps/arena-v2/cli-flag-mapping.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
+# Arena V1 → V2 Flag Audit
+
+This document maps all arena v1 training CLI flags to arena v2 YAML schema coverage.
+
+**Scope:** `arena submit` training subcommands  
+**Source of truth:** arenaa v2 YAML Schema + arena v1 CLI
```
Comment (cheyang) [inline keps/arena-v2/cli-flag-mapping.md]: Typo: `arenaa v2 YAML Schema` → `arena v2 YAML Schema`.

File: keps/arena-v2/README.md [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+**E2E tests:**
+
+- Submit real training jobs to a test cluster
+- Verify CRD structure matches operator expectations
+- Test job lifecycle (submit → running → completed/failed)
+- Validate per-role configuration (worker vs master vs PS)
+
+### Graduation Criteria
+
+**Alpha/MVP (Phase 1 - current):**
```
Comment (cheyang) [inline keps/arena-v2/README.md:188]: Phase 1 says "Core training submission (PyTorchJob, TFJob, MPIJob) with YAML schema" but the sub-bullet only lists `worker` config. That implies TFJob (chief/ps/evaluator) and MPI (launcher) aren't configurable in Phase 1, which contradicts the framework list above. Could you clarify what "Core" actually covers in Phase 1 — single-worker only, or full role coverage?

**Review by cheyang — APPROVED** [review]
/lgtm
/approve
